tests/test_threadsafe.py

```diff
@@ -0,0 +1,219 @@
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+'''
+tests.test_threadsafe
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Unit tests for the :class:`thermoanalysis.ThermoAnalysis` calls are thread-safe
+using the new `thalflex` code
+
+NOTE: It's challenging to get a race condition to compromise thread safety
+as the thal() calls are very fast
+
+'''
+import random
+import threading
+import time
+import unittest
+from typing import (
+    Any,
+    Dict,
+)
+
+from primer3 import thermoanalysis  # type: ignore
+
+
+def make_random_params() -> Dict[str, Any]:
+    '''
+    Returns:
+        dictionary of random parameters for a call  to
+        :meth`ThermoAnalysis.set_thermo_args`
+
+    '''
+    return dict(
+        mv_conc=random.uniform(1, 200),
+        dv_conc=random.uniform(0, 40),
+        dntp_conc=random.uniform(0, 20),
+        dna_conc=random.uniform(0, 200),
+        temp_c=random.randint(10, 70),
+        max_loop=random.randint(10, 30),
+    )
+
+
+def make_rand_seq_pair() -> Dict[str, str]:
+    '''
+    Returns:
+        dictionary of 2 unique sequences keyed by seq1 and seq2
+
+    '''
+    return dict(
+        seq1=''.join([
+            random.choice('ATGC') for _ in
+            range(random.randint(20, 59))
+        ]),
+        seq2=''.join([
+            random.choice('ATGC') for _ in
+            range(random.randint(20, 59))
+        ]),
+    )
+
+
+class TestThermoAnalysisInThread(unittest.TestCase):
+
+    def test_calc_heterodimer_in_thread(self):
+        '''Test basic heterodimer input in a thread
+
+        '''
+        def het_thread(
+                _results_list,
+                i,
+        ):
+            _taf = thermoanalysis.ThermoAnalysis()
+            _params = make_random_params()
+            _taf.set_thermo_args(
+                **_params,
+            )
+            _seq_pair = make_rand_seq_pair()
+            _res = _taf.calc_heterodimer(
+                **_seq_pair,
+            )
+            _results_list[i] = (_res, _params, _seq_pair)
+
+        thread_count = 10
+        results_list = [None] * thread_count
+        thread_list = []
+        t0 = time.time()
+        # 1. Run in threads
+        for i in range(thread_count):
+            t = threading.Thread(
+                target=het_thread,
+                args=(results_list, i),
+            )
+            thread_list.append(t)
+            t.start()
+
+        for t in thread_list:
+            t.join()
+
+        print(f'total: {(time.time()-t0):0.2}')
+
+        taf = thermoanalysis.ThermoAnalysis()
+        t0 = time.time()
+
+        # 2. Compare threaded results to sequential results
+        for res, params, seq_pair in results_list:
+            taf.set_thermo_args(
+                **params,
+            )
+            new_res = taf.calc_heterodimer(
+                **seq_pair,
+            )
+            self.assertAlmostEqual(new_res.tm, res.tm)
+            self.assertAlmostEqual(new_res.dh, res.dh)
```

Comment:
I found that it's reasonable to `assertEqual` in the threaded v/s sequential comparison -- I would expect the floating point output to be identical / deterministic across the two. What do you think? (applies to other cases below as well)

Reply:
Yeah, I did `assertEqual` before and upgraded it to `assertAlmostEqual` as a precaution. 
It seems ok to revert it to `assertEqual`

Reply:
done